src/scheduler/scheduler.rs

```diff
@@ -51,41 +47,46 @@ unsafe impl<VM: VMBinding> Sync for GCWorkScheduler<VM> {}
-            WorkBucketStage::CalculateForwarding => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::SecondRoots => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::RefForwarding => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::FinalizableForwarding => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::Compact => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::Release => WorkBucket::new(false, worker_monitor.clone()),
-            WorkBucketStage::Final => WorkBucket::new(false, worker_monitor.clone()),
+            WorkBucketStage::Unconstrained => WorkBucket::new(true, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Prepare => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Closure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::SoftRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::WeakRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::FinalRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::PhantomRefClosure => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::CalculateForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::SecondRoots => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::RefForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::FinalizableForwarding => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Compact => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Release => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
+            WorkBucketStage::Final => WorkBucket::new(false, worker_monitor.clone(), worker_group.clone()),
         };
 
         // Set the open condition of each bucket.
         {
             // Unconstrained is always open. Prepare will be opened at the beginning of a GC.
             // This vec will grow for each stage we call with open_next()
-            let mut open_stages: Vec<WorkBucketStage> = vec![Unconstrained, Prepare];
+            let first_stw_stage = work_buckets.iter().nth(1).map(|(id, _)| id).unwrap();
+            let mut open_stages: Vec<WorkBucketStage> = vec![first_stw_stage];
             // The rest will open after the previous stage is done.
+            let stages = work_buckets
+                .iter()
+                .map(|(stage, _)| stage)
+                .collect::<Vec<_>>();
```

Comment:
You may use static methods from the `Enum<T>` type in enum-map 0.6.2:
```rust
let first_stw_stage: WorkBucketStage = Enum::<WorkBucketStage>::from_usize(1);

let possible_values: usize = <WorkBucketStage as Enum::<WorkBucketStage>>::POSSIBLE_VALUES;
let stages = (0..possible_values)                                                                   
        .map(|i| {                                                                                      
            let stage: WorkBucketStage = Enum::<WorkBucketStage>::from_usize(i);                        
            stage                                                                                       
        }).collect::<Vec<_>>();                                                                         
```

It is a bit awkward because the `enum-map` crate which mmtk-core currently depends on is **way too old** (0.6.2 vs the latest 2.2.0).

If we update `Cargo.toml` and bump the version to `enum-map = "=2.1.0"`, we will be able to do them much more elegantly:

```rust
let first_stw_stage = WorkBucketStage::from_usize(1);
let stages = (0..WorkBucketStage::LENGTH).map(WorkBucketStage::from_usize);
```

If you feel convenient, you can update the dependency for us in this PR, too.  Note that `2.2.0` requires rustc 1.60.0 which our moma machines currently don't have.  As a workaround, `enum-map = "=2.1.0"` will lock the version to exactly 2.1.0.  We should let our administrator update our installations when appropriate.

Reply:
@wks @qinsoon I think we should leave the code as is for now. I won't be physically in Canberra so I will hold off image updates of CI machines until I get back unless it's urgent (i.e. security vulnerability).

I think this also raises a meta-question of what's our MSRV policy. Currently, our MSRV is 1.59.0, which is not that old. Any change to the MSRV could potentially affect any other projects that depend on us. Of course, we can make the decision that we aggressively update MSRV. But that's something we should at least discuss as a group.

Reply:
Depending on way too old stuff is kinda bad, in my opinion. If we can even just bump to `v2.1.0` without disrupting the MSRV version, I think we should do that. (At some point we should do an audit that updates our dependencies)

Reply:
I switched to `enum-map = "=2.1.0"` and removed the code that fetches the first stage by using `.iter()`.